### ISSUE TITLE
Hash collisions with attribute(name) string fields in ranking expressions

### ISSUE THREAD

**Issue by PeriLara** [vespa-engine/vespa#24898]
Context: 

The vespa environment I'm working on:
OS: Debian
Infra: self hosted
Vespa version: 7.581.33
and 8.5B documents in my index

I have a string field defined as is: 
```
field url_exact type string {
    indexing: summary | attribute
    attribute: fast-search
    match {
        exact
        exact-terminator: "@@"
        cased
    }
}
```

I am trying to evaluate a vespa pipeline from filtering to rank-2. 
For each step I create a vespa query which retrieves relevant surviving documents regarding a ranking step (index coverage, after filtering, after match-phase, after rank-1 and in then, after rank 2).

My question is related to the evaluation of recall at match-phase step, here is what I do to compute it:
```
rank-profile hack_matchPhaseFilter inherits root {
    function relevant_url_filter_scorer() {
       expression: if(attribute(url_exact) == query(relevant_url), 1, 0)
    }
    first-phase {
       expression {
          rankingExpression(relevant_url_filter_scorer)
       }
    }
}
```
This rank profile retrieves an URL from the query and computes an exact match between an input URL and the document URL.

For some URLs, this function returns a match with an other URL. 
E.g: 
The input query: 
```
{
	"searchChain": "vespa_searcher:1.0.2", 
 	"model.queryString": "le monde", 
	"ranking": {
 		"features": {
			"query(relevant_url)": "https://www.lepoint.fr/tags/le-monde"
		},
		"profile": "hack_matchPhaseFilter"
 	},
	"hits": 2,
	"presentation.summary": "mercury_summary"
}
```

The output: 
```
{
   "id": "index:mercury/3/426562517e2ae00b9673a474",
   "relevance": 1.0,
   "source": "mercury",
   "fields": {
	"matchfeatures": {
		"attribute(url_exact)": 2.061907081E9,
                "query(relevant_url)": 2.061907081E9,
                "rankingExpression(relevant_url_filter_scorer)": 1.0
	},
	"sddocname": "mercury",
	"url": "https://letscolourindonesia.com/9/ziyhe/",
   }
},
```

On the other hand, when the input URL is https://letscolourindonesia.com/9/ziyhe/, it returns the relevant URL: 

```
{
   "id": "index:mercury/3/426562517e2ae00b9673a474",
   "relevance": 1.0,
   "source": "mercury",
   "fields": {
	"matchfeatures": {
		"attribute(url_exact)": 2.061907081E9,
                "query(relevant_url)": 2.061907081E9,
                "rankingExpression(relevant_url_filter_scorer)": 1.0
	},
	"sddocname": "mercury",
	"url": "https://letscolourindonesia.com/9/ziyhe/",
   }
},
```

Which would mean that it is du to a hash collision, probably due to the large number of documents that we have, 32 bits hashing might not be enough.  

If it is not clear, I can provide more information, let me know !

Thanks a lot!

**Comment by baldersheim** [comment]
We do indeed use 32 bit hashing.
@bratseth I suspect we can not change this without breaking trained models.
Do we need explicit control of hash function via config override, or do we need explicit syntax in ranking ?

**Comment by bratseth** [comment]
I don't see why we cannot change it by default? It's removing a source of noise.

2^32 bits is insufficient with 8.5B docs if we don't do verification on the retrieved values. So I think we either need that or double the bits, but is either cheap enough to not need config?

**Comment by baldersheim** [comment]
I think the hash function we are using is rather slow. Wonder if it is a variant of the java hash.
I think using XXH3 or similar while be far better and faster.

**Comment by bratseth** [comment]
Yes, and we do have 64 bits available if we just use the bit pattern.
How about doing this also in vespalib string_hash? https://github.com/vespa-engine/vespa/pull/24935

**Comment by jobergum** [comment]
Isn't the hash function used for `attribute(stringfield)` for `==` comparison different than when using string tensor labels which I believe use 64 bits? 

So instead of 

```
field url_exact type string { indexing: attribute} 
```

```
field url_tensor type tensor<int8>(url{}) { indexing: attribute}
```

Then instead of 

```
function relevant_url_filter_scorer() {
       expression: if(attribute(url_exact) == query(relevant_url), 1, 0)
}
```

Use

```
function relevant_url_filter_scorer() {
       expression: sum(query(relevant_url) * attribute(url_tensor))
}
```

Where `query(relevant_url)` is a `<int8>(url{}) ` tensor.

**Comment by bratseth** [comment]
Yes, that should work.
Anyway, the root issue is fixed by moving to 64 bits in https://github.com/vespa-engine/vespa/pull/24938 which is available in 8.88.10 and higher.
